homeassistant/components/tado/coordinator.py

```diff
@@ -0,0 +1,390 @@
+"""Coordinator for the Tado integration."""
+
+from __future__ import annotations
+
+from datetime import datetime, timedelta
+import logging
+from typing import Any
+
+import PyTado
+import PyTado.exceptions
+from PyTado.interface import Tado
+from requests import RequestException
+
+from homeassistant.components.climate import PRESET_AWAY, PRESET_HOME
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryError, HomeAssistantError
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, INSIDE_TEMPERATURE_MEASUREMENT, PRESET_AUTO, TEMP_OFFSET
+
+_LOGGER = logging.getLogger(__name__)
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=4)
+SCAN_INTERVAL = timedelta(minutes=5)
+SCAN_MOBILE_DEVICE_INTERVAL = timedelta(seconds=30)
+
+
+class TadoDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage API calls from and to Tado via PyTado."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        username: str,
+        password: str,
+        fallback: str,
+        debug: bool = False,
+    ) -> None:
+        """Initialize the Tado data update coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=SCAN_INTERVAL,
+        )
+        self.hass = hass
+        self._username = username
+        self._password = password
+        self._fallback = fallback
+        self._debug = debug
+
+        self.tado = None
+        self.home_id: int
+        self.home_name: str
+        self.zones: list[dict[Any, Any]] = []
+        self.devices: list[dict[Any, Any]] = []
+        self.data: dict[str, dict] = {
+            "device": {},
+            "mobile_device": {},
+            "weather": {},
+            "geofence": {},
+            "zone": {},
+        }
+
+    @property
+    def fallback(self):
+        """Return fallback flag to Smart Schedule."""
+        return self._fallback
+
+    async def _async_setup(self):
+        """Set up Tado connection and load initial data."""
+        try:
+            _LOGGER.info("Setting up Tado connection")
```

Comment:
Only have code in the try block that can raise. Would rather use debug for this one btw